From 1660ba0f4de94cfd97a6aec97519e794e24f5ca4 Mon Sep 17 00:00:00 2001
From: Xu Liang <lxu@maxlinear.com>
Date: Wed, 9 Jun 2021 14:33:55 +0800
Subject: [PATCH] PONRTSYS-9067: GSW API - change default configuration for
 extraction

When MPE flag is 10b, disable segmentation to force BM buffer usage.
When Extraction flag is on, redirect packet to PMAC in short cut mode.
---
 .../net/ethernet/lantiq/switch-api/gsw_defconf.c   | 34 +++++++++++++---------
 1 file changed, 20 insertions(+), 14 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
index f204690d97b4..24a5701404ce 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
@@ -90,29 +90,31 @@ int gsw_set_def_pce_qmap(struct core_ops *ops)
 			memset(&q_map, 0, sizeof(GSW_QoS_queuePort_t));
 			q_map.nPortId = gsw_pce_path[j].eg_lpid;
 
-			if ((gswdev->gsw_mode & BIT(0)) == GSW_SHORTCUT_MODE) {
+			q_map.nQueueId = gsw_pce_path[j].qid;
+			q_map.nRedirectPortId =	gsw_pce_path[j].redir_lpid;
+
+			q_map.nTrafficClassId = i;
+
+			if (gsw_pce_path[j].ext == X) {
+				q_map.bExtrationEnable = 1;
+				ops->gsw_qos_ops.QoS_QueuePortSet(ops, &q_map);
+
+				q_map.bExtrationEnable = 0;
+			} else {
+				q_map.bExtrationEnable = gsw_pce_path[j].ext;
+			}
+
+			if (!q_map.bExtrationEnable &&
+			    (gswdev->gsw_mode & BIT(0)) == GSW_SHORTCUT_MODE) {
 				if (q_map.nPortId == LOG_3) {
 					q_map.nQueueId = 0;	/* Use Q0 */
 					q_map.nRedirectPortId = LOG_3;
 				} else if (q_map.nPortId == LOG_4) {
 					q_map.nQueueId = 8;	/* Use Q8 */
 					q_map.nRedirectPortId = LOG_4;
-				} else {
-					q_map.nQueueId = gsw_pce_path[j].qid;
-					q_map.nRedirectPortId =
-						gsw_pce_path[j].redir_lpid;
 				}
-			} else {
-				q_map.nQueueId = gsw_pce_path[j].qid;
-				q_map.nRedirectPortId =
-					gsw_pce_path[j].redir_lpid;
 			}
 
-			if (gsw_pce_path[j].ext != X)
-				q_map.bExtrationEnable = gsw_pce_path[j].ext;
-
-			q_map.nTrafficClassId = i;
-
 			ops->gsw_qos_ops.QoS_QueuePortSet(ops, &q_map);
 		}
 	}
@@ -524,6 +526,10 @@ static int pmac_eg_cfg(struct core_ops *ops, u8 pmacid, u8 dpu)
 					eg_cfg.nFlowIDMsb	= k;
 					eg_cfg.bFcsEna		= 1;
 
+					/* MPE flag (10b) is extraction to CPU */
+					if ((j & 3) == 2)
+						eg_cfg.bBslSegmentDisable = 1;
+
 					/* All other fields set to 0. */
 					ops->gsw_pmac_ops.Pmac_Eg_CfgSet(ops,
 									 &eg_cfg);
